src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/Tracker.cs

```diff
@@ -3,83 +3,136 @@
-            return GetEntry(item).Ambiguous;
+            return TryGetEntry(item)?.Ambiguous ?? false;
         }
 
         public void MarkAmbiguous(GraphItem<TItem> item)
         {
-            GetEntry(item).Ambiguous = true;
+            GetOrAddEntry(item).Ambiguous = true;
         }
 
         public bool IsBestVersion(GraphItem<TItem> item)
         {
-            var entry = GetEntry(item);
-
-            var version = item.Key.Version;
+            var entry = TryGetEntry(item);
 
-            foreach (var known in entry.List)
+            if (entry is not null)
             {
-                if (version < known.Key.Version)
+                var version = item.Key.Version;
+
+                foreach (var known in entry.Items)
                 {
-                    return false;
+                    if (version < known.Key.Version)
+                    {
+                        return false;
+                    }
                 }
             }
 
             return true;
         }
 
-        public IEnumerable<GraphItem<TItem>> GetDisputes(GraphItem<TItem> item) => GetEntry(item).List;
+        public IEnumerable<GraphItem<TItem>> GetDisputes(GraphItem<TItem> item)
+        {
+            return TryGetEntry(item)?.Items ?? Enumerable.Empty<GraphItem<TItem>>();
+        }
 
         internal void Clear()
         {
-            _entries.Clear();
+            _entryByLibraryName.Clear();
         }
 
-        private Entry GetEntry(GraphItem<TItem> item)
+        private Entry? TryGetEntry(GraphItem<TItem> item)
         {
-            Entry itemList;
-            if (!_entries.TryGetValue(item.Key.Name, out itemList))
+            _entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry);
+            return entry;
+        }
+
+        private Entry GetOrAddEntry(GraphItem<TItem> item)
+        {
+            if (!_entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry))
             {
-                itemList = new Entry();
-                _entries[item.Key.Name] = itemList;
+                entry = new Entry();
+                _entryByLibraryName[item.Key.Name] = entry;
             }
-            return itemList;
+
+            return entry;
         }
 
-        private class Entry
+        private sealed class Entry
         {
-            public Entry()
+            /// <summary>
+            /// This field can have one of three values:
+            ///
+            /// - null: when no graph items exist in this entry
+            /// - a graph item: when only a single graph item exists in this entry
+            /// - a hash set of graph items: when multiple graph items exist in this entry
+            ///
+            /// This packing exists in order to reduce the size of this object in
+            /// memory. For large graphs, this can amount to a significant saving in memory,
+            /// which helps overall performance.
+            /// </summary>
+            private object? _storage;
```

Comment:
Given casting has a perf cost, plus `object` obscures types, is it worthwhile having two fields instead, `private TItem? _single` and `private HashSet<TItem>? _collection` instead?  It gives more compile time static type analysis.

Reply:
The goal here was to reduce memory footprint, and a second field would increase size again. Type checking is not free, but it's still very fast. The scenario where this was flagged as contributing to GC, the CPU was spending a huge amount of time on GC. My sense is that reducing GC pressure would offset any overhead from type checking. To write a benchmark you'd need realistic data loads, which would take time to gather. I don't think it'd be worth the effort to investigate given how much low hanging fruit exists throughout NuGet.

Reply:
I agree with Drew's assessment here.